server/src/main/java/org/apache/cloudstack/vm/UnmanagedVMsManagerImpl.java

```diff
@@ -512,11 +519,21 @@ private Pair<UnmanagedInstanceTO.Disk, List<UnmanagedInstanceTO.Disk>> getRootAn
                 rootDisk = disk;
             } else {
                 dataDisks.add(disk);
+                DiskOffering diskOffering = diskOfferingDao.findById(dataDiskOfferingMap.getOrDefault(disk.getDiskId(), null));
+                if ((disk.getCapacity() == null || disk.getCapacity() <= 0) && diskOffering != null) {
+                    disk.setCapacity(diskOffering.getDiskSize());
+                }
             }
         }
-        if (diskIdsWithoutOffering.size() > 1) {
-            throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, String.format("VM has total %d disks, disk offering mapping not provided for %d disks. Disk IDs that may need a disk offering - %s", disks.size(), diskIdsWithoutOffering.size()-1, String.join(", ", diskIdsWithoutOffering)));
+        if (diskIdsWithoutOffering.size() > 1 || rootDisk == null) {
+            throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, String.format("VM has total %d disks, disk offering mapping not provided for %d disks. Disk IDs that may need a disk offering - %s", disks.size(), diskIdsWithoutOffering.size() - 1, String.join(", ", diskIdsWithoutOffering)));
```

Comment:
In which case rootDisk is null at this point when we have already checked `callerDiskIds.size() != disks.size() - 1`?